src/api/uploads/MultiputPart.js

```diff
@@ -26,38 +30,289 @@ class MultiputPart extends Base {
         this.numDigestRetriesPerformed = 0;
         this.numUploadRetriesPerformed = 0;
         this.offset = offset;
-        this.sha256 = null;
         this.size = size;
         this.state = PART_STATE_NOT_STARTED;
         this.timing = {};
         this.uploadedBytes = 0;
+        this.data = {};
+        this.uploadUrl = `${this.uploadHost}/api/2.0/files/upload_sessions/${sessionId}`;
+        this.config = config;
+        this.rangeEnd = offset + size - 1;
+        this.onSuccess = onSuccess || noop;
+        this.onError = onError || noop;
+        this.onProgress = onProgress || noop;
+        this.getPartsState = getPartsState;
+        this.getNumPartsUploading = getNumPartsUploading;
     }
 
+    toJSON = () =>
+        JSON.stringify({
+            index: this.index,
+            offset: this.offset,
+            size: this.size,
+            state: this.state,
+            uploadedBytes: this.uploadedBytes,
+            numUploadRetriesPerformed: this.numUploadRetriesPerformed,
+            numDigestRetriesPerformed: this.numDigestRetriesPerformed,
+            sha256: this.sha256,
+            timing: this.timing
+        });
+
+    /**
+     * Returns file part associated with this Part.
+     *
+     * @return {Object}
+     */
+    getPart = (): Object => this.data.part;
+
+    /**
+     * Uploads this Part via the API. Will retry on network failures.
+     *
+     * @return {void}
+     */
+    upload = (): void => {
+        if (this.isDestroyed() || !this.sha256 || !this.blob) {
+            return;
+        }
+
+        const clientEventInfo = {
+            documentHidden: document.hidden,
+            digest_retries: this.numDigestRetriesPerformed,
+            timing: this.timing,
+            parts_uploading: this.getNumPartsUploading()
+        };
+
+        const headers = {
+            'Content-Type': 'application/octet-stream',
+            Digest: `sha-256=${this.sha256}`,
+            'Content-Range': `bytes ${this.offset}-${this.rangeEnd}/${this.size}`,
+            'X-Box-Client-Event-Info': JSON.stringify(clientEventInfo)
+        };
+
+        this.startTimestamp = Date.now();
+
+        this.xhr.uploadFile({
+            url: this.uploadUrl,
+            data: this.blob,
+            headers,
+            method: 'PUT',
+            successHandler: this.uploadSuccessHandler,
+            errorHandler: this.uploadErrorHandler,
+            progressHandler: this.uploadProgressHandler,
+            withIdleTimeout: true,
+            idleTimeoutDuration: this.config.requestTimeoutMS
+        });
+    };
+
+    /**
+     * Handler for upload part success
+     * 
+     * @param {Object} data
+     * @return {void}
+     */
+    uploadSuccessHandler = (data: Object) => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.state = PART_STATE_UPLOADED;
+        this.consoleLogFunc(() => `Upload completed: ${this.toJSON()}. Parts state: ${this.getPartsState()}`);
```

Comment:
If you need to pass in the getPartsState function into this class just for this log line, it may not be worth it.

Reply:
You mean we don't need to log the parts states here? Or are you suggesting some other approach to make an individual part aware of the big picture?

Reply:
Definitely not needed, except for potential debugging purposes in the future.  If you want to simplify the code to get rid of it, I'd say go for it